src/test/IO/test_IO_DiscretisedDensity.cxx

```diff
@@ -0,0 +1,146 @@
+//
+//
+/*!
+
+  \file
+  \ingroup IO
+
+  \brief A simple program to test the stir::OutputFileFormat function.
+
+  \author Kris Thielemans
+  \author Richard Brown
+
+
+
+  To run the test, you should use a command line argument with the name of a file.
+  This should contain a test par file.
+  See stir::OutputFileFormatTests class documentation for file contents.
+
+  \warning Overwrites files STIRtmp.* in the current directory
+
+  \todo The current implementation requires that the output file format as also
+  readable by stir::read_from_file. At least we should provide a
+  run-time switch to not run that part of the tests.
+*/
+/*
+    Copyright (C) 2002- 2011, Hammersmith Imanet Ltd
+    Copyright (C) 2018- , University College London
+
+    This file is part of STIR.
+
+    This file is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    This file is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    See STIR/LICENSE.txt for details
+*/
+
+#include "stir/IO/test/test_IO.h"
+
+#ifndef STIR_NO_NAMESPACES
+using std::cerr;
+using std::endl;
+using std::ifstream;
+using std::istream;
+#endif
+
+START_NAMESPACE_STIR
+
+/*!
+  \ingroup test
+  \brief A simple class to test the OutputFileFormat function.
+
+  The class reads input from a stream, whose contents should be as
+  follows:
+
+  \verbatim
+  Test OutputFileFormat Parameters:=
+  output file format type :=
+  ; here are parameters specific for the file format
+  End:=
+  \endverbatim
+
+  \warning Overwrites files STIRtmp.* in the current directory
+  \todo Delete STIRtmp.* files, but that's a bit difficult as we don't know which ones
+  are written.
+*/
+class IOTests_DiscretisedDensity : public IOTests<DiscretisedDensity<3,float> >
+{
+public:
+    explicit IOTests_DiscretisedDensity(istream& in) : IOTests(in) {}
+
+protected:
+
+    void create_image();
+    void read_image();
+    void check_result();
+};
+
+void IOTests_DiscretisedDensity::create_image()
+{
+    _image_to_write_sptr = create_single_image();
+}
+
+void IOTests_DiscretisedDensity::read_image()
+{
+    // now read it back
+    unique_ptr<DiscretisedDensity<3,float> >
+        density_ptr = read_from_file<DiscretisedDensity<3,float> >(_filename);
+
+    if(!check(!is_null_ptr(density_ptr), "failed reading"))
+        return;
+
+    _image_to_read_sptr.reset(
+                new VoxelsOnCartesianGrid<float>(
+                    *dynamic_cast< VoxelsOnCartesianGrid<float> *>(density_ptr.get())));
```

Comment:
I (think) have seen this syntax in some places. Why not use the get_empty_copy() ??